phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala

```diff
@@ -0,0 +1,269 @@
+/*                     __                                                   *\
+**     ________ ___   / /  ___      __ ____  PhantomJS support for Scala.js **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2017, LAMP/EPFL       **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    https://www.scala-js.org/      **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                                   **
+**                          |/____/                                         **
+\*                                                                          */
+
+package org.scalajs.jsenv.phantomjs
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Promise, Future}
+import scala.util.control.NonFatal
+
+import java.io._
+import java.net._
+import java.nio.file.{Files, StandardCopyOption}
+
+import org.scalajs.jsenv._
+
+import org.scalajs.io._
+import org.scalajs.io.URIUtils.fixFileURI
+import org.scalajs.io.JSUtils.escapeJS
+
+private final class ComRun(jettyClassLoader: ClassLoader,
+    runConfig: RunConfig, onMessage: String => Unit,
+    startRun: VirtualBinaryFile => JSRun)
+    extends JSComRun {
+
+  import ComRun._
+
+  import runConfig.logger
+
+  private val promise = Promise[Unit]()
+
+  def future: Future[Unit] = promise.future
+
+  private val sendBuffer = mutable.ListBuffer.empty[String]
+  private var canSendNow: Boolean = false
+
+  private def loadMgr(): WebsocketManager = {
+    val loader =
+      if (jettyClassLoader != null) jettyClassLoader
+      else getClass().getClassLoader()
+
+    val clazz = loader.loadClass(
+        "org.scalajs.jsenv.phantomjs.JettyWebsocketManager")
+
+    val ctors = clazz.getConstructors()
+    assert(ctors.length == 1, "JettyWebsocketManager may only have one ctor")
+
+    val listener = new WebsocketListener {
+      def onRunning(): Unit = startPhantomJSProcess()
+      def onOpen(): Unit = sendPendingMessages()
```

Comment:
@gzm0 said:

> Please abstract this over Futures so there is less callback. It might be a good idea to adapt our shim for classloading to give access to futures (and only take an `onMessage` and `log` handle).